README
Devel-Kit version 0.7

DOCUMENTATION

See POD for documentation.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

Alternatively, to install with Module::Build, you can use the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install

DEPENDENCIES

See DEPENDENCIES section in POD, 'requires' key in Build.PL, 
or 'PREREQ_PM' key in Makefile.PL

COPYRIGHT AND LICENCE

Copyright (C) 2012, Daniel Muey

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

Note to self about building:

    perltidy -b t/00.load.t t/01.import.t t/02.functions.t
    rm -f t/00.load.t.bak t/01.import.t.bak t/02.functions.t.bak

    cp t/00.load.t t/00.load-TAP.t
    cp t/00.load.t t/00.load-cPanel.t

    cp t/01.import.t t/01.import-TAP.t
    cp t/01.import.t t/01.import-cPanel.t

    perl -pi -e 's{Devel::Kit}{Devel::Kit::TAP}g'  t/00.load-TAP.t  t/01.import-TAP.t
    perl -pi -e 's{Devel::Kit}{Devel::Kit::cPanel}g'  t/00.load-cPanel.t  t/01.import-cPanel.t

    perl -pi -e 's{(use\s*Test::More.*)}{$1\nBEGIN { eval "require Cpanel::Logger;";plan skip_all => "tests irrelevant on non-cPanel environment" if \$@; };\n}g'  t/00.load-cPanel.t  t/01.import-cPanel.t


Changes
Revision history for Devel-Kit

0.7  Sun Dec  1 20:38:42 2013
    - add a()
    - improve punycode logic
    - add si()

0.6  Wed Sep  4 20:47:04 2013
    - add Hex support via xe() and xu()

0.5  Tue Jul 16 21:34:25 2013
    - require min version of module want
    - Add _CODE.pm to manifest
    - not use 'Devel::Kit' in ni() test to avoid confusion
    
0.4  Sun Jul 14 23:13:37 2013
    - Add coderef stat dumping: ci()
    - Add name space dumping support: ni()
        - strict/warnings import flag changed to 'no' so it doesn't conflict w/ ni()
    - add devel info (made ri() for refs and added to vd(ARG,1) for strings)
    - add Regex debugging: rx()
    - add misc runtime environment dumping: ei()
    - implement se() and su()
    - Add tidy test

0.3  Thu Nov 22 10:28:29 2012
    - replace temp functions with module
    - rt 80378: Add handy alias command to POD
    - rt 79991: catch File::Slurp failures and report error instead of dying
    - Add crockford base32 encode/decode

0.2  Sat Aug 25 10:33:03 CEST 2012
    - use more universally installable punycode module
    - tests for @ behaviors
    - rt 79159: POD fixes, thanks Kevin!
    - Sub class tests (minor snafu on that and I wanted it out initially during YAPC::EU 2012) README has notes-to-self
    - adjust test regexp to try and address erroneous failure on cpantesters.org 4cb7b3be-6f98-1014-af33-fcab386ac21f

0.1  Thu May  3 21:28:17 2012
       Initial release (@ YAPC::EU Wed Aug 22 15:26:18 CEST 2012)


Build.PL
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    module_name       => 'Devel::Kit',
    license           => 'perl',
    dist_author       => 'Daniel Muey <http://drmuey.com/cpan_contact.pl>',
    dist_version_from => 'lib/Devel/Kit.pm',
    requires          => {
        'Test::More'      => 0,    # for testing, I know I know: Test::XT
        'File::Temp'      => 0,    # for testing, I know I know: Test::XT
        'Test::Mock::Cmd' => 0,    # for testing, I know I know: Test::XT

        'Module::Want'        => 0.5,
        'String::UnicodeUTF8' => 0,

        # lazy loaded (via Module::Want) only when needed
        'Import::Into'        => 0,
        'Data::Dumper'        => 0,
        'File::Slurp'         => 0,
        'YAML::Syck'          => 0,,
        'JSON::Syck'          => 0,,
        'XML::Parser'         => 0,
        'Storable'            => 0,
        'Config::INI::Reader' => 0,
        'Data::MessagePack'   => 0,

        'Digest::MD5' => 0,
        'Digest::SHA' => 0,

        'MIME::QuotedPrint'          => 0,
        'Net::IDN::Encode'           => 0,
        'HTML::Entities'             => 0,
        'URI::Escape'                => 0,
        'MIME::Base64'               => 0,
        'Convert::Base32::Crockford' => 0,

        'Devel::Symdump'  => 0,
        'Time::HiRes'     => 0,
        'Unix::PID::Tiny' => 0,
        'Devel::CountOps' => 0,

        'Devel::Size'   => 0,
        'Devel::Peek'   => 0,
        'Capture::Tiny' => 0,

        'Regexp::Debugger' => 0,

        'App::Kit' => 0.6,

        'IPC::Open3::Utils' => 0.91,
    },
    add_to_cleanup => ['Devel-Kit-*'],
);

$builder->create_build_script();


lib/Devel/Kit/cPanel.pm
package Devel::Kit::cPanel;

use strict;
use warnings;
$Devel::Kit::cPanel::VERSION = '0.1';

use base 'Devel::Kit';
@Devel::Kit::cPanel::ISA = ('Devel::Kit');

sub import {
    my $caller = caller();
    require Import::Into;    # die here since we don't need it otherwise, so we know right off there's a problem, and so caller does not have to check status unless they want to
    Devel::Kit->import::into( $caller, @_ );
}

use Cpanel::Logger ();
my $logger = Cpanel::Logger->new();

my $a;
{
    no warnings 'redefine';
    *Devel::Kit::o = sub {
        unshift @_, $logger;
        goto &Cpanel::Logger::info;
    };

    *Devel::Kit::a = sub {
        if ( !$ak ) {
            require App::Kit::cPanel;
            $ak = App::Kit::cPanel->instance;
        }
        return $ak;
    };
}

1;

__END__

=encoding utf-8

=head1 NAME

Devel::Kit::cPanel - cPanel friendly version of Devel::Kit (A handy toolbox of things to ease development/debugging).

=head1 VERSION

This document describes Devel::Kit::cPanel version 0.1

=head1 SYNOPSIS

    use Devel::Kit::cPanel; # strict and warnings are now enabled
    
    d($something); # d() and some other useful debug/dump functions are now availble!

=head1 DESCRIPTION

cPanel friendly (via Cpanel::Logger->info()) version of L<Devel::Kit>.

See L<Devel::Kit> for details.

=head1 INCOMPATIBILITIES

None reported.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests to
C<bug-devel-kit@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org>.

=head1 AUTHOR

Daniel Muey  C<< <http://drmuey.com/cpan_contact.pl> >>

=head1 LICENCE AND COPYRIGHT

Copyright (c) 2012, Daniel Muey C<< <http://drmuey.com/cpan_contact.pl> >>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.

=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.


Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME          => 'Devel::Kit',
    AUTHOR        => 'Daniel Muey <http://drmuey.com/cpan_contact.pl>',
    VERSION_FROM  => 'lib/Devel/Kit.pm',
    ABSTRACT_FROM => 'lib/Devel/Kit.pm',
    PL_FILES      => {},
    LICENSE       => 'perl',
    PREREQ_PM     => {
        'Test::More'      => 0,    # for testing, I know I know: Test::XT
        'File::Temp'      => 0,    # for testing, I know I know: Test::XT
        'Test::Mock::Cmd' => 0,    # for testing, I know I know: Test::XT

        'Module::Want'        => 0.5,
        'String::UnicodeUTF8' => 0,

        # lazy loaded (via Module::Want) only when needed
        'Import::Into' => 0,

        'Data::Dumper'        => 0,
        'File::Slurp'         => 0,
        'YAML::Syck'          => 0,,
        'JSON::Syck'          => 0,,
        'XML::Parser'         => 0,,
        'Storable'            => 0,
        'Config::INI::Reader' => 0,
        'Data::MessagePack'   => 0,

        'Digest::MD5' => 0,
        'Digest::SHA' => 0,

        'MIME::QuotedPrint'          => 0,
        'Net::IDN::Encode'           => 0,
        'HTML::Entities'             => 0,
        'URI::Escape'                => 0,
        'MIME::Base64'               => 0,
        'Convert::Base32::Crockford' => 0,

        'Devel::Symdump'  => 0,
        'Time::HiRes'     => 0,
        'Unix::PID::Tiny' => 0,
        'Devel::CountOps' => 0,

        'Devel::Size'   => 0,
        'Devel::Peek'   => 0,
        'Capture::Tiny' => 0,

        'Regexp::Debugger' => 0,

        'App::Kit' => 0.6,

        'IPC::Open3::Utils' => 0.91,
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES    => 'Devel-Kit-*' },
);
